### PR TITLE
Changelog for master

### PR DESCRIPTION
**Pull request by Lex-2008**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @olehermanse can review this?

**Comment by olehermanse** [comment]
@Lex-2008 Why 2? https://github.com/cfengine/core/pull/3415

**Comment by Lex-2008** [comment]
that one was submitted in error

File: ChangeLog [hunk 1]
```diff
@@ -1,3 +1,162 @@
...
+	  feature_tls_1_2                 source=agent,hardclass
+	  feature_tls_1_3                 source=agent,hardclass
+	- Add a new variable $(sys.default_policy_path)
+	  A new sys variable that provides the path of the default policy
+	  file evaluated when no file is specified with the '-f' option.
+	- Add an option to skip the initial policy run on bootstrap
+	  In some cases it may not be desired to run the policy as the last
+	  step of the bootstrap. This can be done with the new
+	  '--skip-bootstrap-policy-run' option for cf-agent.
+	- Add commit title to changelog by default, if it has ticket number
```
Comment (vpodzime) [inline ChangeLog]: Drop

File: ChangeLog [hunk 2]
```diff
@@ -1,3 +1,162 @@
...
+	  for example to do health checks of databases on startup.
+	  Ticket: (ENT-4064)
+	- Added function string_replace. (CFE-2850)
+	- Added special behavior for bundles named __main__
+	  If the bundle is defined in the entry policy it will be defined as main.
+	  If the bundle is defined elsewhere, it will be removed.
+	  This makes it easy to make importable library policy which can also
+	  be executed directly.
+	- Added tests for function string_replace (CFE-2850)
+	- Added tests for stack data structure. (ENT-3830)
```
Comment (vpodzime) [inline ChangeLog]: drop entries for tests

File: ChangeLog [hunk 3]
```diff
@@ -1,3 +1,162 @@
...
+	- Added thread-safe stack data structure. (ENT-3830)
```
Comment (vpodzime) [inline ChangeLog]: internal detail

File: ChangeLog [hunk 4]
```diff
@@ -1,3 +1,162 @@
...
+	- Allow dots in variable identifiers with no such bundle
+	  As described and discussed in , defining remote variables
```
Comment (vpodzime) [inline ChangeLog]: missing link?

File: ChangeLog [hunk 5]
```diff
@@ -1,3 +1,162 @@
...
+	  (injecting variables into remote bundles) is dangerous and must
+	  be blocked. However, using a dot-separated common prefix for
+	  variables raises no security concerns and can be considered
+	  valid. (CFE-1915)
+	- Allow requiring TLS 1.3 as the minimum version
+	- Allow up to 5 file descriptors per connection in cf-serverd
```
Comment (vpodzime) [inline ChangeLog]: drop

File: ChangeLog [hunk 6]
```diff
@@ -1,3 +1,162 @@
...
+	  (CFE-2575)
+	- Apply augments after vars, classes and inputs in def.json
+	  (CFE-2741, CFE-2844)
+	- Avoid using GetDefaultBinDir() on Windows (ENT-4115)
```
Comment (vpodzime) [inline ChangeLog]: drop

File: ChangeLog [hunk 7]
```diff
@@ -1,3 +1,162 @@
...
+	- Fixed memory leak in JSON to policy conversion (ENT-4136)
+	- Fixed memory leak in lmdb cleanup (CFE-2918)
+	- Fixed memory leaks in cf-agent during bootstrap (CFE-2918)
+	- Fixed memory leaks in variablesmatching() and findfiles() (CFE-2918)
+	- Fixed missing class with mustache templates in warn_only mode
+	  (CFE-2600)
+	- Fixed small memory leak in cf-serverd (CFE-2918)
+	- Fixed small memory leak in cf-upgrade (ENT-4136)
+	- Fixed small memory leaks of environment variable strings (CFE-2918)
+	- Fixed valgrind warning about lmdb writing uninitialized memory
```
Comment (vpodzime) [inline ChangeLog]: drop, the entries about memory leaks are useful, but this one is not

File: ChangeLog [hunk 8]
```diff
@@ -1,3 +1,162 @@
...
+	              meta => {"os[linux]"};
+	          "matches"
+	              slist => variablesmatching(".*", "os\[linux\]");
+	      reports:
+	          "Match: $(matches)";
+	  }
+	  The above example is correct. If you don't escape the brackets
+	  like above, it will no longer work. (You probably shouldn't use
+	  brackets in tags anyway).
+	- Masterfiles stage readme was changed. (ENT-3806)
```
Comment (vpodzime) [inline ChangeLog]: drop

File: ChangeLog [hunk 9]
```diff
@@ -1,3 +1,162 @@
...
+	- Prevent the init script from managing processes inside containers
+	  (ENT-3800)
+	- Read mustache-rendered files in text mode when comparing digest
+	  (ENT-2526)
+	- Reload persistent classes on config reload in cf-execd and cf-serverd
+	  (CFE-2857)
+	- Test that classes can be defined without constraints (CFE-2000)
+	- Trigger promises.cf as the last step of bootstrap (CFE-2826)
+	- We must start on line position is 1 instead of 0 to support the @if
```
Comment (vpodzime) [inline ChangeLog]: confusing

File: ChangeLog [hunk 10]
```diff
@@ -1,3 +1,162 @@
...
+	  macros on line 1 in a file.
+	  Ticket: (CFE-2817)
+	- cf-agent intermittently hangs on windows (ENT-3756)
```
Comment (vpodzime) [inline ChangeLog]: looks like a new feature

File: ChangeLog [hunk 11]
```diff
@@ -1,3 +1,162 @@
...
+	- cf-check: Added diagnose command to cf-check (ENT-4064)
+	- change GIT_BRANCH to GIT_REFSPEC and remove Design Center vars
+	  (ENT-4023)
+	- change code coverage destination to work with atom IDE plugin. Should work
```
Comment (vpodzime) [inline ChangeLog]: drop

File: ChangeLog [hunk 12]
```diff
@@ -1,3 +1,162 @@
...
+	  fine with codecov.io (ENT-3756)
+	- disable/enable tests according to their behaviour on new buildslaves
```
Comment (vpodzime) [inline ChangeLog]: drop

File: ChangeLog [hunk 13]
```diff
@@ -1,3 +1,147 @@
...
+	  like cf-agent would hang.
+	  cf-check replaces our lmdb database dumper, lmdump.
+	  `cf-check lmdump` or symlinking / renaming it to lmdump
+	  will make cf-check have the exact same behavior as lmdump.
+	  cf-check will include much more functionality in the future
+	  and some of the code will be added to other binaries,
+	  for example to do health checks of databases on startup.
+	  Ticket: (ENT-4064)
+	- Added function string_replace. (CFE-2850)
+	- Added special behavior for bundles named __main__
```
Comment (olehermanse) [inline ChangeLog]: This feature was part of 3.12.0 and shouldn't be mentioned here.

File: ChangeLog [hunk 14]
```diff
@@ -1,3 +1,147 @@
...
+	  (CFE-2793)
+	- Class names set by module protocol are automatically canonified
+	  (CFE-2877, CFE-2887)
+	- Classes failsafe_fallback and bootstrap_mode are now reported by default
+	- Correct log level for data_readstringarray* (CFE-2922)
+	- Do not iterate over JSON objects' properties in mustache (CFE-2125)
+	- Do not render templates when passed invalid data (CFE-2194)
+	- Eliminated error messages caused by attempting to kill expired processes
+	  (CFE-2824)
+	- Extend the test to cover more cases (ENT-3959)
```
Comment (olehermanse) [inline ChangeLog]: Doesn't make sense witout further detail.

File: ChangeLog [hunk 15]
```diff
@@ -1,3 +1,147 @@
...
+	  The above example is correct. If you don't escape the brackets
+	  like above, it will no longer work. (You probably shouldn't use
+	  brackets in tags anyway).
+	- Prevent the init script from managing processes inside containers
+	  (ENT-3800)
+	- Read mustache-rendered files in text mode when comparing digest
+	  (ENT-2526)
+	- Reload persistent classes on config reload in cf-execd and cf-serverd
+	  (CFE-2857)
+	- Test that classes can be defined without constraints (CFE-2000)
```
Comment (olehermanse) [inline ChangeLog]: This one can probably be removed.

File: ChangeLog [hunk 16]
```diff
@@ -1,3 +1,147 @@
...
+	- Trigger promises.cf as the last step of bootstrap (CFE-2826)
```
Comment (olehermanse) [inline ChangeLog]: Move this up to the entry about `--skip-bootstrap-policy-run` which is referencing this feature.

File: ChangeLog [hunk 17]
```diff
@@ -1,3 +1,147 @@
...
+	- Fixed issue with @if macro failing when it is on the first line.
+	  (CFE-2817)
+	- Fixed issue with cf-agent intermittently hanging on windows
+	  sometimes (ENT-3756)
+	- cf-check: Added diagnose command to cf-check (ENT-4064)
```
Comment (olehermanse) [inline ChangeLog]: Remove this entry.

**Comment by Lex-2008** [comment]
thanks, fix'd!

File: ChangeLog [hunk 18]
```diff
@@ -1,3 +1,139 @@
...
+	- Do not render templates when passed invalid data (CFE-2194)
+	- Eliminated error messages caused by attempting to kill expired processes
+	  (CFE-2824)
+	- Fix cf-runalerts systemd unit conditions so the service will run
+	  (ENT-3929)
+	- Fix the off-by-one error in cf-runagent background process spawning
+	  (CFE-2873)
+	- Fixed OOB read / heap buffer overflow in evaluator (ENT-4136)
+	- Fixed a memory leak which occured when reloading RSA keys from disk
+	- Fixed a memory leak which occured while loading augments files
```
Comment (craigcomstock) [inline ChangeLog]: Did this RSA keys entry need a JIRA ticket reference?

File: ChangeLog [hunk 19]
```diff
@@ -1,3 +1,139 @@
...
+	  (ENT-2526)
+	- Reload persistent classes on config reload in cf-execd and cf-serverd
+	  (CFE-2857)
+	- Fixed issue with @if macro failing when it is on the first line.
+	  (CFE-2817)
+	- Fixed issue with cf-agent intermittently hanging on windows
+	  sometimes (ENT-3756)
+	- change GIT_BRANCH to GIT_REFSPEC and remove Design Center vars
+	  (ENT-4023)
+	- missed cleanup.h include due to only building --enable-debug
```
Comment (craigcomstock) [inline ChangeLog]: missed cleanup.h doesn't seem like a significant change release entry.

File: ChangeLog [hunk 20]
```diff
@@ -1,3 +1,139 @@
...
+	  	ID=coreos
+	  	VERSION_ID=1185.3.0
+	  For the example above, `sys.flavor` will be `coreos_1185` and 4 hard
+	  classes will be set; `coreos_1185_3_0`, `coreos_1185_3`, `coreos_1185`,
+	  and `coreos`.
+	  For backwards compatibility, older distribution specific logic is still
+	  executed and may overwrite sys.flavor and define hard classes as before.
+	- refactor use of atexit to use custom cleanup function instead. On Windows
+	  atexit() unloads DLLs before and/or during atexit functions being called
+	  which cases bad behavior. (ENT-3756)
```
Comment (craigcomstock) [inline ChangeLog]: s/which cases/which causes/

File: ChangeLog (hunk 18 above)
Comment (Lex-2008) [reply ChangeLog]: it came from this commit:
commit https://github.com/cfengine/core/commit/a03e82ba369b2fb3612a21d51d118aaad5417b74
Author: @olehermanse 
Date:   Tue Oct 30 14:28:33 2018 +0100

    Fixed a memory leak which occured when reloading RSA keys from disk
    
    Changelog: Title

from this PR: https://github.com/cfengine/core/pull/3396

So it should be CFE-2857, indeed

File: ChangeLog (hunk 20 above)
Comment (Lex-2008) [reply ChangeLog]: good eyes! :eyes:

**Pull request merged**